src/book/02-projects/01-nubots/03-build-system.mdx

```diff
@@ -5,3 +5,526 @@ title: Build System
+A list of the currently used compiler flags for the `nuc7i7bnh` can be seen in [generate_nuc7i7bnh_toolchain.py](https://github.com/NUbots/NUbots/blob/master/docker/usr/local/toolchain/generate_nuc7i7bnh_toolchain.py).
+
+If you are setting up a new toolchain targetting a robot that is using a different CPU to the `nuc7i7bnh`, log on to the new robot, install `gcc` (`sudo pacman -S gcc`) and execute the following commands
+
+```bash
+target=$(gcc -march=native -Q --help=target | grep -- '-march=' | cut -f3 | head -n1)
+diff -y --suppress-common-lines <(gcc -march=native -Q --help=target) <(gcc -march=${target} -Q --help=target)
+```
+
+The first command will determine the codename of the CPU family (i.e. for an Intel CPU it might be `broadwell`, `haswell`, `skylake`, etc).
+
+For the rest of this example we will assume that `target=skylake`.
+
+The second command compares the compiler flags that compiling with `-march=native` would enable compared to using `-march=skylake`.
+The output of this function could look like this
+
+```bash
+-mabm        [enabled]        |    -mabm        [disabled]
+-mrtm        [enabled]        |    -mrtm        [disabled]
+```
+
+The left column are instructions that are enabled/disabled when compiling using `-march=native` and the right column are the instructions that are enabled/disabled when compiling using `-march=skylake`.
+Combining both columns tells us that to get from a generic `skylake` CPU to the CPU that we are using we need to enable both the `abm` and the `rtm` instructions using `-mabm` and `-mrtm`.
+
+<Alert type='info'>
+
+If the output was actually
+
+```bash
+-mabm        [disabled]       |    -mabm        [enabled]
+-mrtm        [enabled]        |    -mrtm        [disabled]
+```
+
+Then we would need to disable `abm` using `-mno-abm` and enable `rtm` using `-mrtm`.
+
+</Alert>
+
+The final piece of information we need is about the CPU cache sizes
+
+```ssh
+gcc -### -E - -march=native 2>&1 | sed -r '/cc1/!d;s/(")|(^.* - )//g'| grep -oP -- "--param l[12]-cache(-line|)-size=[0-9]*"
+```
+
+The output of this command will look like this
+
+```bash
+--param l1-cache-size=32
+--param l1-cache-line-size=64
+--param l2-cache-size=12288
+```
+
+<Alert type='warning'>
+
+If you get an error saying "zsh: bad pattern: -###", wrap double-quotes around the `-###` in the above command.
+
+</Alert>
+
+Putting all of this together, the `targets` dict for our example target would be
+
+```python
+target = {
+    "flags": [
+        "-march=skylake",
+        "-mtune=skylake",
+        "-mabm",
+        "-mrtm",
+        "--param l1-cache-size=32",
+        "--param l1-cache-line-size=64",
+        "--param l2-cache-size=12288",
+        "-fPIC",
+    ],
+    "release_flags": ["-O3", "-DNDEBUG"],
+    "asm_flags": ["-DELF", "-D__x86_64__", "-DPIC"],
+    "asm_object": "elf64",
+}
```

Comment:
For some reason, the syntax highlighting here looks terrible, maybe try showing it as javascript and see if that looks better?

Reply:
Just had a look at [GitHub's Markdown render](https://github.com/NUbots/NUbook/blob/biddulph/build-system/src/book/02-projects/01-nubots/03-build-system.mdx#the-toolchain) for reference and it's not much better.

Maybe JS will look better, but it feels weird changing the language for syntax highlighting.

Reply:
I know, but it’s either that or fixing GitHub’s syntax highlighter.

Reply:
I removed the syntax highlighting. Nothing I tried looked particularly nice for this